### PR TITLE
All councils page.

### PR DESCRIPTION
**Pull request by whoojemaflip**
A sortable, filterable index page to browse all local authorities. 

https://trello.com/c/6xQryalX/537-update-view-all-councils-page-5

llm-all-councils-implementation

Missing from the design is the fancy-pants search/filtering widget. There was some feedback from the user tests that I think we wanted to discuss before investing too much time in a build.  This solution uses the Table Filter from `govuk_admin_template` which does the basics for us.  A more sophisticated thing could be built on top of this, once we're sure what to go for.

### CODE REVIEW

File: db/migrate/20161104150651_add_broken_link_count_to_local_authority.rb [hunk 1]
```diff
@@ -0,0 +1,11 @@
+class AddBrokenLinkCountToLocalAuthority < ActiveRecord::Migration[5.0]
+  def up
+    add_column :local_authorities, :broken_link_count, :integer, default: 0
+
+    LocalAuthority.all.map(&:calculate_count_of_broken_links)
```
Comment (davidbasalla) [inline db/migrate/20161104150651_add_broken_link_count_to_local_authority.rb:5]: Quick question, is this doing the same as `LocalAuthority.all.each { |la| la.calculate_count_of_broken_links }`? Does using the `map` method have any advantages here?

Comment (whoojemaflip) [reply db/migrate/20161104150651_add_broken_link_count_to_local_authority.rb:5]: It's exactly the same functionally, but I feel the shorter syntax is a little more intention revealing. It can be a bit confusing the first time you see it though, I agree. Would you rather it were swapped out for the long-hand?

File: app/models/local_authority.rb [hunk 2]
```diff
@@ -14,6 +14,13 @@ def provided_services
     Service.for_tier(self.tier).enabled
   end
 
+  def calculate_count_of_broken_links
+    # to ensure that we can replay our migrations, wrap the update in a guard clause.
+    if respond_to?(:broken_link_count)
+      update_attribute(:broken_link_count, self.links.where.not(status: 200).count)
```
Comment (davidbasalla) [inline app/models/local_authority.rb]: We were wondering if broken links == links.where.not(status: 200) and if there were any other status codes that might count as 'broken', such as redirects. But looking at the link checker it looks like redirects do return a 200 (if final redirects return a 200) so this looks fine. Briefly thought about defining a `broken` scope on the Link model to move the definition of what broken links are to the Link Model... not sure if necessary at this point

File: spec/models/link_spec.rb [hunk 3]
```diff
@@ -155,4 +155,13 @@
     expect(@link.status).to be nil
     expect(@link.link_last_checked).to be nil
   end
+
+  describe "after_update" do
+    it "triggers a recount of the local authority's broken links" do
```
Comment (davidbasalla) [inline spec/models/link_spec.rb]: Would it be worth adding another test here to demonstrate that `calculate_count_of_broken_links` will not get called if `status` does not change?

File: spec/models/local_authority_spec.rb [hunk 4]
```diff
@@ -76,5 +76,13 @@
         expect(@local_authority.link_last_checked).to be_nil
       end
     end
+
+    describe "calculate_count_of_broken_links" do
```
Comment (davidbasalla) [inline spec/models/local_authority_spec.rb]: Small style niggle, I'm used to seeing method tests have a `#` in front of them, see http://betterspecs.org/#describe

Comment (davidbasalla) [reply spec/models/local_authority_spec.rb]: Also the word 'links' is missing from the test description

Comment (davidbasalla) [reply spec/models/local_authority_spec.rb]: We could add another link here with status 200 to show that the broken link count is correct in counting only links that are broken.

File: app/models/local_authority.rb (hunk 2 above)
Comment (whoojemaflip) [reply app/models/local_authority.rb]: I really like that. It makes complete sense that when we talk about Broken Links that they be either a subset of all links aka. `Link.all.broken` as you suggest or even a first class model object eg. `class BrokenLink < Link`

File: spec/models/link_spec.rb (hunk 3 above)
Comment (whoojemaflip) [reply spec/models/link_spec.rb]: 👍

File: spec/models/local_authority_spec.rb (hunk 4 above)
Comment (whoojemaflip) [reply spec/models/local_authority_spec.rb]: well spotted. #willfix

Comment (davidbasalla) [reply spec/models/local_authority_spec.rb]: Were also thinking that we wouldn't need to define a local authority as you could use the `local_authority` association from the link factory, but maybe it's more readable? _shrug_

File: db/migrate/20161104150651_add_broken_link_count_to_local_authority.rb (hunk 1 above)
Comment (davidbasalla) [reply db/migrate/20161104150651_add_broken_link_count_to_local_authority.rb:5]: No it's good like this, was just not used to seeing it.

File: app/views/local_authorities/index.html.erb [hunk 5]
```diff
@@ -10,7 +10,9 @@
   <table class="table table-striped table-hover table-bordered contacts-table" data-module="filterable-table">
     <thead>
       <tr class="table-header">
-        <th>Local Authorities (<%= @authorities.count %>)</th>
+        <th>
+          <div class='count'><span><%= @authorities.count %></span> local authorities</div>
```
Comment (davidbasalla) [inline app/views/local_authorities/index.html.erb]: Why did you add a "span" element here? Is that good practice generally?

Comment (whoojemaflip) [reply app/views/local_authorities/index.html.erb]: I think github ate your markup.. which element are you referring to?

Comment (davidbasalla) [reply app/views/local_authorities/index.html.erb]: ^ thanks, fixed :) 

So there's a spec which tests the wording that this commit changes. It gets updated in [another commit in this PR](https://github.com/alphagov/local-links-manager/pull/71/commits/c40c642b7615bee8347b3e93521bc662dd37536b). Might be worth either pulling that test change into this commit, or rolling this commit into the other one...

File: spec/support/appear_before_matcher.rb [hunk 6]
```diff
@@ -0,0 +1,32 @@
+module SortOrderMatchers
```
Comment (davidbasalla) [inline spec/support/appear_before_matcher.rb:1]: This is cool! Did you get this pattern from somewhere or is it completely hand rolled? I vaguely remember testing sort order can be a pain...

File: spec/factories/links.rb [hunk 7]
```diff
@@ -5,6 +5,10 @@
     url { "http://#{local_authority.slug}.example.com/#{service_interaction.service.slug}/#{service_interaction.interaction.label.parameterize}" }
     status nil
     link_last_checked nil
+
+    trait :with_service_interaction do
```
Comment (davidbasalla) [inline spec/factories/links.rb]: This trait is probably not needed because we've already defined the service_interaction association for this factory. We tried removing this trait and any calls to it and the tests still passed.

File: spec/models/local_authority_spec.rb [hunk 8]
```diff
@@ -76,5 +76,13 @@
...
+      it "calculates the correct number of broken for the local authority" do
+        local_authority = FactoryGirl.create(:local_authority)
+        FactoryGirl.create(:link, :with_service_interaction, local_authority: local_authority, status: 500)
+        expect { local_authority.calculate_count_of_broken_links }.to change { local_authority.broken_link_count }.from(0).to(1)
```
Comment (davidbasalla) [inline spec/models/local_authority_spec.rb]: We can't actually get this test to pass. It appears that when the link instance is being created by FactoryGirl, the after_save callback is already being called so when we get to L84 the broken_link_count is already 1, not 0 as currently expected...

File: app/views/local_authorities/index.html.erb [hunk 9]
```diff
@@ -19,14 +19,13 @@
...
-      locals: {placeholder: "Filter list of local authorities"} %>
     </thead>
     <tbody>
       <% @authorities.each do |authority| %>
         <tr>
-          <td class="name">
-            <%= link_to authority.name, local_authority_services_path(authority.slug), class: 'js-open-on-submit' %>
+          <td>
+            <%= link_to authority.name, local_authority_services_path(authority.slug), class: 'js-open-on-submit local-authority-name' %>
+            <div class="count"><span><%= authority.broken_link_count %> </span>Broken links</div>
```
Comment (davidbasalla) [inline app/views/local_authorities/index.html.erb]: Could we add a test for this to the integration specs? Not sure how fiddly testing the rows would be...

**Comment by davidbasalla** [comment]
This looks good, we've finished reviewing the commits and left notes accordingly. I didn't look at the CSS too much as it's really my 'fortay' but I'm sure it's good! :)

File: app/views/local_authorities/index.html.erb (hunk 5 above)
Comment (whoojemaflip) [reply app/views/local_authorities/index.html.erb]: the `span` element here gives us the ability to target the numeric bit of the sentence for fancy CSS formatting.

File: spec/support/appear_before_matcher.rb (hunk 6 above)
Comment (whoojemaflip) [reply spec/support/appear_before_matcher.rb:1]: I went looking for it on the internet, and rolled my own based on someone else's code which did 200% of what we needed here. Custom Matchers are the best.

**Comment by whoojemaflip** [comment]
@davidbasalla @emmabeynon That was a thorough review! Thanks for pushing me to do better :)

I've rebased the suggested changes into the relevant commits to preserve the history.

As discussed, there's some light refactoring done to the `LinkStatusUpdater` class which allows us to simplify the `Link` and `LocalAuthority` classes from the initial design. Due to a race-hazard, it was a bunch more efficient to drop that refactoring into this PR rather than create several more blocker dependencies.

File: lib/local-links-manager/check_links/link_status_updater.rb [hunk 10]
```diff
@@ -3,29 +3,27 @@
 module LocalLinksManager
   module CheckLinks
     class LinkStatusUpdater
```
Comment (emmabeynon) [inline lib/local-links-manager/check_links/link_status_updater.rb:5]: This looks good.  We should probably change similar code in the HomepageStatusUpdater class too so that it's consistent.

Comment (emmabeynon) [reply lib/local-links-manager/check_links/link_status_updater.rb:5]: Small change should be made to the commit message so it starts with an active verb.

File: app/models/link.rb [hunk 11]
```diff
@@ -19,6 +19,11 @@ class Link < ApplicationRecord
       .where(service_interactions: { service_id: service })
   }
 
+  HTTP_OK_STATUS_CODE = 200
+
+  scope :currently_broken, -> { where.not(status: HTTP_OK_STATUS_CODE) }
```
Comment (emmabeynon) [inline app/models/link.rb:23]: Would it be worth having tests for these new scopes?

File: app/models/link.rb [hunk 12]
```diff
@@ -1,6 +1,5 @@
 class Link < ApplicationRecord
   before_update :set_time_and_status_on_updated_link, if: :url_changed?
-  after_save :trigger_recount_of_broken_links, if: :status_changed?
```
Comment (emmabeynon) [inline app/models/link.rb]: Something we just thought of is that when an individual link is updated, without this `after_save` callback the broken link recount will not occur, and therefore the broken link count will be incorrect.  We should add some tests for this and figure out how to best trigger this recount for individually edited links.

File: app/models/link.rb (hunk 11 above)
Comment (whoojemaflip) [reply app/models/link.rb:23]: The code is run by the integration tests, so it seemed like overkill to write slow database-bound unit tests for something like this.

Additionally, unit testing a simple scope like this seems a bit like writing tests for ActiveRecord.  What do you think? I'm comfortable with the current level of coverage for this.

Comment (davidbasalla) [reply app/models/link.rb:23]: Sounds good, agreed that the scope is fairly simple so not necessary to test further

File: lib/local-links-manager/check_links/link_status_updater.rb [hunk 13]
```diff
@@ -8,15 +8,29 @@ def initialize(link_checker = LinkChecker.new)
       end
 
       def update
+        check_all_links
+        update_local_authorities_with_broken_link_count
```
Comment (davidbasalla) [inline lib/local-links-manager/check_links/link_status_updater.rb]: Just thinking about this... currently the broken link counts for LAs only get updated after all links have been checked. Currently this is run at 2am every night and takes about 6.5 hours to finish (estimate based on highly unscientific guesswork), so the broken link counts would only be updated around 8:30 am. Are we happy with this sort of delay in the results? Anyone using the service before 8:30 am would see wrong results...

File: lib/local-links-manager/check_links/link_status_updater.rb (hunk 10 above)
Comment (whoojemaflip) [reply lib/local-links-manager/check_links/link_status_updater.rb:5]: I've just looked at the HomepageStatusUpdater class for the first time. It looks like the updater was maybe planned to be abstracted into common library code at some point, but that hadn't happened yet. I'm not sure what the benefits of consistency are, but I agree that it will need a look at some point.

What with the plans to build a more encompassing Link Checker, do you agree that we should leave this code as-is, for now at least?

File: app/models/link.rb (hunk 12 above)
Comment (whoojemaflip) [reply app/models/link.rb]: good catch

File: lib/local-links-manager/check_links/link_status_updater.rb (hunk 13 above)
Comment (whoojemaflip) [reply lib/local-links-manager/check_links/link_status_updater.rb]: maybe we should recalculate after every link check completes?

File: lib/local-links-manager/check_links/link_status_updater.rb (hunk 10 above)
Comment (brenetic) [reply lib/local-links-manager/check_links/link_status_updater.rb:5]: The one con of that is that the link checker is months away from discovery let alone actually building it. It will come after Lolotron V2 and mid (maybe after, depending . . . ) migration

**Comment by brenetic** [comment]
👍
